Fix gc hole in runtime/gc_ctrl.c

DemiMarie opened the pull request:
If the evaluation of one of the arguments to caml_alloc_3 triggers a garbage collection, the other arguments will not be valid anymore.

This rule was not documented in the OCaml manual, so I added an explanation of the problem.

gasche reviewed (commented):
This is a good catch, thanks!

I'm not fond of the proposed fixes.

1. I think that changing the manual would need substantially more work. The style is inconsistent with the rest of the manual (you use concepts like "to hole" that are not used anywhere else), and I don't think that adding these as extra gc rules fits the purpose and scope of the rest of this manual section, which generally assumes that all temporaries are properly registered as roots. A minimal fix would be would to add an explicit comment to Rule 2 that temporaries are equivalent to local variables but cannot be declared with `CAMLlocal`, so temporaries of `value` type must be written to registered locals right away to "live in harmony with the GC".

2. The code fix feels too clever and annoyingly asymmetric to me. I would be in favor of sticking to the boring style where we register everything, unless there are good performance reasons to do anything more complex. In this case I don't think there are reasons to do complex  (obviously the performance of this code could be improved by returning a record of (unboxed) floats instead of a tuple, but no users expressed such a need so far), so I think you should do *really* simple and have three locals.

DemiMarie committed:
runtime/gc_ctrl.c: Fix registration of temporaries with GC

If the evaluation of one of the arguments to caml_alloc_3 triggers a
garbage collection, the other arguments will not be valid anymore.

gasche commented on manual/src/cmds/intf-c.etex:
```diff
@@ -999,7 +999,10 @@ Primitives (C functions that can be called from OCaml) should never return void.
+function, not in a nested block.  Temporaries are equivalent to
+local variables, but they cannot be registered with the garbage
+collector, so a temporary of type "value" must not be live when
+a garbage colleciton may occur.
```
```suggestion
a garbage collection may occur.
```

gasche commented on runtime/gc_ctrl.c:
```diff
@@ -103,20 +103,17 @@ CAMLprim value caml_gc_minor_words(value v)
+  Store_double_flat_field(v, 0, minwords);
+  Store_double_flat_field(v, 1, prowords);
+  Store_double_flat_field(v, 2, majwords);
+  return v;
```
I don't think that this is correct, because you are changing the return type without changing the OCaml consumers. It would be much simpler (in terms of reasoning about backward compatibility) to keep returning a tuple of boxed floats.

gasche commented on runtime/gc_ctrl.c:
```diff
@@ -103,20 +103,23 @@ CAMLprim value caml_gc_minor_words(value v)
+  minwords_ = caml_copy_double(minwords);
+  prowords_ = caml_copy_double(prowords);
+  majwords_ = caml_copy_double(majwords);
+  v = caml_alloc_small(3, 0);
```
I would have used `value res = ...` instead of weirdly reusing the input variable, and then `caml_alloc_3` as before, but you've suffered enough roundtrips from me by now (today is the day I discovered the document listing [review antipatterns](https://www.chiark.greenend.org.uk/~sgtatham/quasiblog/code-review-antipatterns/) and I feel personally targeted by this list!) and I'm fine with the current version.

DemiMarie replied:
Thanks!  I’d argue that `caml_alloc_N` functions are themselves antipatterns, as they must pointlessly reregister their arguments.  That’s why I went with `caml_alloc_small`.  Yes, users might well choose to trade performance for convenience, but I feel that the OCaml distribution itself should not.

gasche replied:
Can you actually measure a performance difference between the two options in a microbenchmark calling `Gc.counters`?

DemiMarie replied:
Probably not.

gasche replied:
I nerd-sniped myself and did the measurements. Here is the benchmark:

```ocaml
let n =
  try int_of_string Sys.argv.(1)
  with _ -> failwith "expected an integer argument on the command line"

let () =
  for _ = 1 to n do
    ignore (Gc.counters ())
  done
```

And here are the versions compared:

```c
/* version: "trunk".
  This version is incorrect as the caml_copy_double() temporaries may get invalidated. */
CAMLprim value caml_gc_counters(value v)
{
  CAMLparam0 ();   /* v is ignored */
  CAMLlocal1 (res);

  /* get a copy of these before allocating anything... */
  double minwords = caml_gc_minor_words_unboxed();
  double prowords = (double)Caml_state->stat_promoted_words;
  double majwords = Caml_state->stat_major_words +
                    (double) Caml_state->allocated_words;

  res = caml_alloc_3(0,
    caml_copy_double (minwords),
    caml_copy_double (prowords),
    caml_copy_double (majwords));
  CAMLreturn (res);
}

/* version "demi" */
CAMLprim value caml_gc_counters(value v)
{
  CAMLparam0 ();   /* v is ignored */
  CAMLlocal3 (minwords_, prowords_, majwords_);

  /* get a copy of these before allocating anything... */
  double minwords = caml_gc_minor_words_unboxed();
  double prowords = (double)Caml_state->stat_promoted_words;
  double majwords = Caml_state->stat_major_words +
                    (double) Caml_state->allocated_words;

  minwords_ = caml_copy_double(minwords);
  prowords_ = caml_copy_double(prowords);
  majwords_ = caml_copy_double(majwords);
  value res = caml_alloc_small(3, 0);
  Field (res, 0) = minwords_;
  Field (res, 1) = prowords_;
  Field (res, 2) = majwords_;
  CAMLreturn (res);
}

/* version "store_field": following the style around in the file */
CAMLprim value caml_gc_counters(value v)
{
  CAMLparam0 ();   /* v is ignored */
  CAMLlocal3 (minwords_, prowords_, majwords_);

  /* get a copy of these before allocating anything... */
  double minwords = caml_gc_minor_words_unboxed();
  double prowords = (double)Caml_state->stat_promoted_words;
  double majwords = Caml_state->stat_major_words +
                    (double) Caml_state->allocated_words;

  minwords_ = caml_copy_double(minwords);
  prowords_ = caml_copy_double(prowords);
  majwords_ = caml_copy_double(majwords);
  value res = caml_alloc_small(3, 0);
  Store_field (res, 0, minwords_);
  Store_field (res, 1, prowords_);
  Store_field (res, 2, majwords_);
  CAMLreturn (res);
}


/* version "clean": the simplest possible */
CAMLprim value caml_gc_counters(value v)
{
  CAMLparam0 ();   /* v is ignored */
  CAMLlocal4 (minwords_, prowords_, majwords_, res);

  /* get a copy of these before allocating anything... */
  double minwords = caml_gc_minor_words_unboxed();
  double prowords = (double)Caml_state->stat_promoted_words;
  double majwords = Caml_state->stat_major_words +
                    (double) Caml_state->allocated_words;

  minwords_ = caml_copy_double(minwords);
  prowords_ = caml_copy_double(prowords);
  majwords_ = caml_copy_double(majwords);
  res = caml_alloc_3(0, minwords_, prowords_, majwords_);
  CAMLreturn (res);
}
```

And the benchmark results:

| Command       |    Mean [ms] | Min [ms] |    Relative |
|:--------------|-------------:|---------:|------------:|
| `trunk`       |  243.3 ± 2.9 |    238.7 |        1.00 |
| `clean`       | 274.1 ± 19.2 |    262.7 | 1.13 ± 0.08 |
| `demi`        | 286.7 ± 26.3 |    267.8 | 1.18 ± 0.11 |
| `store_field` | 359.3 ± 25.2 |    341.6 | 1.48 ± 0.11 |

It turns out that the `clean` version I was suggesting is in fact slightly faster than your version (in ways that probably don't matter in practice).

DemiMarie replied:
Good catch!  I’m curious why that is, but it’s definitely the way to go.

gasche replied:
I also looked and I'm not sure. The code of `caml_alloc_3` is careful to only register more roots *if* a collection is needed, so in fact there is no additional rooting overhead. I played a bit with godbolt to understand why the version was slightly faster instead of same-speed, and I noticed differences in spilling (maybe spilling double temporaries is more costly), but nothing that clearly explains the small gap.

gasche reviewed (approved):
The documentation extension is fine and the new code is correct yet simple. Approved. Thanks!

gasche committed:
nitpick the gc_counters code

jmid commented:
@Octachron If you are rolling a 5.2.1 release, consider cherry-picking this bugfix PR too :pray:

DemiMarie commented:
> @Octachron If you are rolling a 5.2.1 release, consider cherry-picking this bugfix PR too 🙏

Also for the 4.x LTS release if that is still supported.

Octachron commented:
Cherry-picked to 5.2 as 57736d16

@DemiMarie , as far as I can see the 4.14 version does not have this issue, since
```
  res = caml_alloc_tuple (3);
  Store_field (res, 0, caml_copy_double (minwords));
  Store_field (res, 1, caml_copy_double (prowords));
  Store_field (res, 2, caml_copy_double (majwords));
  CAMLreturn (res);
  ```
  stores every temporary in `res` as soon as it is allocated.
